File: packages/wio_terminal/go/wifi/wifi.go
// Package wifi is the TinyGo runtime half of the dart2tinygo `wio_terminal`
// Wi-Fi binding (Dart side: `lib/wifi.dart`, imported as `wiowifi`). Kept as
// its own Go sub-package (still inside the wio_terminal/go module) so an
// example that doesn't use Wi-Fi doesn't pull `net/http` and the RTL8720DN
// driver into its build; see `docs/writing_bindings.md`.
package wifi

import (
	"io"
	"net/http"
	"strings"

	"tinygo.org/x/drivers/netdev"
	"tinygo.org/x/drivers/netlink"
	"tinygo.org/x/drivers/netlink/probe"
)

// WiFi is the Wio Terminal's Wi-Fi module (RTL8720DN), for station-mode
// connections and simple HTTP requests. Requires RTL8720DN firmware 2.1.2
// or later; see https://wiki.seeedstudio.com/Wio-Terminal-Network-Overview/.
type WiFi struct {
	link      netlink.Netlinker
	dev       netdev.Netdever
	connected bool
}

// NewWiFi probes and initializes the RTL8720DN. Does not connect to an
// access point yet; call Connect.
func NewWiFi() *WiFi {
	link, dev := probe.Probe()
	return &WiFi{link: link, dev: dev}
}

// Connect joins the given access point. Returns whether it succeeded
// (the annotation machinery has no way to surface the underlying Go error).
func (w *WiFi) Connect(ssid, password string) bool {
	err := w.link.NetConnect(&netlink.ConnectParams{
		Ssid:       ssid,
		Passphrase: password,
	})
	w.connected = err == nil
	return w.connected
}

// IsConnected reports whether the last Connect call succeeded and
// Disconnect hasn't been called since.
func (w *WiFi) IsConnected() bool {
	return w.connected
}

// IPAddress returns the assigned IPv4 address, or "" if not connected.
func (w *WiFi) IPAddress() string {
	addr, err := w.dev.Addr()
	if err != nil {
		return ""
	}
	return addr.String()
}

// Disconnect leaves the access point.
func (w *WiFi) Disconnect() {
	w.link.NetDisconnect()
	w.connected = false
}

// HttpGet fetches url and returns the response body as text, or "" on
// failure (network error, non-readable body, ...).
func (w *WiFi) HttpGet(url string) string {
	resp, err := http.Get(url)
	if err != nil {
		return ""
	}
	defer resp.Body.Close()
	body, err := io.ReadAll(resp.Body)
	if err != nil {
		return ""
	}
	return string(body)
}

// HttpPost posts body to url with the given content type and returns the
// response body as text, or "" on failure.
func (w *WiFi) HttpPost(url, contentType, body string) string {
	resp, err := http.Post(url, contentType, strings.NewReader(body))
	if err != nil {
		return ""
	}
	defer resp.Body.Close()
	data, err := io.ReadAll(resp.Body)
	if err != nil {
		return ""
	}
	return string(data)
}

File: packages/wio_terminal/lib/wifi.dart
/// dart2tinygo binding for the Wio Terminal's Wi-Fi module (RTL8720DN).
///
/// A separate library (and Go sub-package, `go/wifi`) from
/// `wio_terminal.dart` so that a program that doesn't `import` this file
/// doesn't pull `net/http` and the RTL8720DN driver into its build; see
/// `docs/writing_bindings.md`.
///
/// Requires RTL8720DN firmware 2.1.2 or later; see
/// https://wiki.seeedstudio.com/Wio-Terminal-Network-Overview/. Building a
/// program that uses `httpGet`/`httpPost` needs a larger goroutine stack
/// than TinyGo's default — pass `-stack-size=4KB` to the `tinygo build` /
/// `tinygo flash` step.
@GoImport('github.com/o-ga09/dart2tinygo/packages/wio_terminal/go/wifi',
    alias: 'wiowifi')
library;

import 'package:tinygo_annotations/tinygo_annotations.dart';

/// Probes and initializes the RTL8720DN. Does not connect to an access
/// point yet; call [WiFi.connect].
@GoName('wiowifi.NewWiFi')
external WiFi newWiFi();

/// The Wio Terminal's Wi-Fi module. Obtain one with [newWiFi].
@GoType('*wiowifi.WiFi')
class WiFi {
  WiFi._();

  /// Joins the given access point. Returns whether it succeeded.
  @GoName('Connect')
  external bool connect(String ssid, String password);

  /// Whether the last [connect] call succeeded and [disconnect] hasn't
  /// been called since.
  @GoName('IsConnected')
  external bool isConnected();

  /// The assigned IPv4 address, or `''` if not connected.
  @GoName('IPAddress')
  external String ipAddress();

  /// Leaves the access point.
  @GoName('Disconnect')
  external void disconnect();

  /// Fetches [url] and returns the response body as text, or `''` on
  /// failure.
  @GoName('HttpGet')
  external String httpGet(String url);

  /// Posts [body] to [url] with [contentType] and returns the response
  /// body as text, or `''` on failure.
  @GoName('HttpPost')
  external String httpPost(String url, String contentType, String body);
}
